src/header.rs

```diff
@@ -0,0 +1,295 @@
+}
+
+impl<A> AsRef<A> for Jwe<A> {
+    fn as_ref(&self) -> &A {
+        &self.additional
+    }
+}
+impl<A> AsMut<A> for Jwe<A> {
+    fn as_mut(&mut self) -> &mut A {
+        &mut self.additional
+    }
+}
+/// Header parameters that are specific to signatures
+#[derive(Debug, Hash, PartialEq, Eq, Serialize, Deserialize, Clone)]
+#[non_exhaustive]
+pub struct Jws<A> {
+    /// `b64` parameter as defined by RFC 7797. This parameter is optional and
+    /// it's default value is `true`.
+    ///
+    /// If this value is `false`, the payload of the JWS is not base64 urlsafe
+    /// encoded. This can work for simple stuff like a hex string, but will
+    /// often cause parsing errors. Use of this option makes sense if the
+    /// payload of a JWS is detached.
+    ///
+    /// Note: In a JsonWebToken, this value MUST always be true, e.g. the
+    /// payload MUST NOT use the unencoded payload option.
+    #[serde(rename = "b64", default = "returns_true")]
+    payload_base64_url_encoded: bool,
+    #[serde(flatten)]
+    additional: A,
+}
+
+impl<A> AsRef<A> for Jws<A> {
+    fn as_ref(&self) -> &A {
+        &self.additional
+    }
+}
+impl<A> AsMut<A> for Jws<A> {
+    fn as_mut(&mut self) -> &mut A {
+        &mut self.additional
+    }
+}
+
+/// A [`JoseHeader`] used with Json Web Signatures
+pub type JwsHeader<T, A> = JoseHeader<T, Jws<A>>;
+/// A [`JoseHeader`] used with Json Web Encryption
+pub type JweHeader<T, A> = JoseHeader<T, Jwe<A>>;
+
+// general implementation for protected and unprotected headers in both jwe and
+// jws
+impl<T, A> JoseHeader<T, A>
+where
+    T: HeaderMarker,
+{
+    /// Convert this [`JoseHeader`] back to a [`JoseHeaderBuilder`]
+    pub fn to_builder(self) -> JoseHeaderBuilder<T, A> {
+        JoseHeaderBuilder::from(self)
+    }
+
+    /// The [algorithm](JsonWebAlgorithm) used in this JWS or JWE
+    pub fn algorithm(&self) -> JsonWebAlgorithm {
+        self.algorithm
+    }
+    // TODO: add other getters
+}
+
+impl<A> JoseHeader<Protected, A> {
+    // implementation for protected headers in both jwe and jws
+}
+
+impl<A> JoseHeader<Unprotected, A> {
+    // implementation for unprotected headers in both jwe and jws
+}
+
+impl<T, A> JwsHeader<T, A>
+where
+    T: HeaderMarker,
+{
+    // implementation for protected and unprotected headers in jws
+}
+
+impl<A> JwsHeader<Protected, A> {
+    // implementation for protected headers in jws
+}
+
+impl<A> JwsHeader<Unprotected, A> {
+    // implementation for unprotected headers in jws
+}
+
+impl<T, A> JweHeader<T, A>
+where
+    T: HeaderMarker,
+{
+    // implementation for protected and unprotected headers in jwe
+}
+
+impl<A> JweHeader<Protected, A> {
+    // implementation for protected headers in jwe
+}
+
+impl<A> JweHeader<Unprotected, A> {
+    // implementation for unprotected headers in jwe
+}
+
+impl<T, A> AsRef<A> for JoseHeader<T, A> {
+    fn as_ref(&self) -> &A {
+        &self.additional
+    }
+}
+impl<T, A> AsMut<A> for JoseHeader<T, A> {
```

Comment:
If we are going to implement `AsMut`, note that the header loses its immutability since the user can change the additional field.